CARDS-814: Add a `flatten` json processor for forms

sofi2002sofi opened the pull request:
[CARDS-814](https://phenotips.atlassian.net/browse/CARDS-814)

Testing notes:
Check the form link with URL suffix .flatten.-identify.deep.json.  For example http://localhost:8080/Forms/84c7c113-4d7e-4b88-b890-4b20240abacf.flatten.-identify.deep.json.

sofi2002sofi committed:
CARDS-814: Add a `flatten` json processor for forms

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,123 @@
+        this.jsonSectionMap.remove();
+    }
+
+    private void addAnswers(final Node node, final JsonObjectBuilder json)
```
Even though private methods are not required to have javadoc, it's useful to add an explanatory comment about why this method exists.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,123 @@
+                final Node child = children.nextNode();
+                final String childId = child.getIdentifier();
+                if (child.isNodeType("cards:Answer") && this.childrenJsons.get().containsKey(childId)) {
+                    final String answerId = childId.substring(childId.length() - 36);
```
No magic numbers, please :)

sofi2002sofi committed:
CARDS-814: Add a flatten json processor for forms

fix childId evaluation

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,128 @@
+            while (children.hasNext()) {
+                final Node child = children.nextNode();
+                final String childId = child.getIdentifier();
+                final String idWithoutParent = childId.substring(childId.length() - ID_LENGTH);
```
It's not a good practice to assume a certain length for identifiers. Instead, we can rely on the known structure of the string and pull out the last segment:
```suggestion
                final String idWithoutParent = childId.substring(childId.lastIndexOf("/") + 1);
```

IntegralProgrammer replied:
A better way to do this is to slice by the `/` character and pick the last element from the list. There are cases where selecting the last 36 characters does not make sense. For example, if there are non-referenceable nodes which do not contain a `jcr:uuid` then `getIdentifier()` will return the node's name in place of a`jcr:uuid`. Therefore, if we were to have a path such as `/12345678-1234-1234-1234-123456789abc/condition/operandB`, slicing out the last 36 characters would result in `1234-123456789abc/condition/operandB` which is an invalid path.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    @Override
+    public JsonValue processChild(Node node, Node child, JsonValue input, Function<Node, JsonValue> serializeNode)
+    {
+        if (input == null) {
+            return null;
+        }
+        try {
+            if (child.isNodeType("cards:Answer")) {
+                this.childrenJsons.get().put(child.getIdentifier(), input.asJsonObject());
+                return null;
+            } else if (child.isNodeType("cards:AnswerSection")) {
+                return null;
+            }
+        } catch (RepositoryException e) {
+            // Shouldn't happen
+        }
+        return input;
+    }
```
This should be documented with what this method is doing. From what I understand by looking at it, each time it is invoked, it is passed a JCR node and if that JCR node is a `cards:Answer` node, this node is serialized and stored to a (`this.childrenJsons`) `ThreadLocal` object. If that is the case, please document it accordingly. Also, please document why `null` is returned if `child` is of type `cards:Answer` or `cards:AnswerSection` but otherwise the `input` `JsonValue` is simply passed-through.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,128 @@
+{
+    private static final int ID_LENGTH = 36;
+
+    private ThreadLocal<Map<String, JsonObject>> childrenJsons = ThreadLocal.withInitial(HashMap::new);
+    private ThreadLocal<JsonObjectBuilder> jsonSection = ThreadLocal.withInitial(Json::createObjectBuilder);
+    private ThreadLocal<Map<String, JsonObjectBuilder>> jsonSectionMap = ThreadLocal.withInitial(HashMap::new);
```
Add comments to describe what information each one of these objects holds.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,128 @@
+                final String idWithoutParent = childId.substring(childId.length() - ID_LENGTH);
+                if (child.isNodeType("cards:Answer") && this.childrenJsons.get().containsKey(childId)) {
+                    final JsonObject childJson = this.childrenJsons.get().get(childId);
+                    if (child.getParent().isNodeType("cards:AnswerSection")) {
```
Do we need to call `child.getParent()`? Can we not simply use `node` as we already know that `child` is a child node of `node`?

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,128 @@
+                    final JsonObject childJson = this.childrenJsons.get().get(childId);
+                    if (child.getParent().isNodeType("cards:AnswerSection")) {
+                        this.jsonSection.get().add(idWithoutParent, childJson);
+                        this.jsonSectionMap.get().put(child.getParent().getIdentifier(), this.jsonSection.get());
```
Do we need to call `child.getParent()`? Can we not simply use `node` as we already know that `child` is a child node of `node`?

IntegralProgrammer commented:
I tested this with a custom made questionnaire with an `Answer` placed 4 `AnswerSection` levels deep and the answer was not available in `.flatten.deep.json`. We should create a _CARDS-814_ test questionnaire and add it to the set of _test questionnaires_ that are available in the `--test` mode. That would allow for very thorough testing of this PR.

IntegralProgrammer reviewed (changes requested):
I tested this with a custom made questionnaire with an `Answer` placed 4 `AnswerSection` levels deep and the answer was not available in `.flatten.deep.json`. We should create a _CARDS-814_ test questionnaire and add it to the set of _test questionnaires_ that are available in the `--test` mode. That would allow for very thorough testing of this PR.

IntegralProgrammer committed:
CARDS-814: Add a flatten json processor for forms

sofi2002sofi committed:
CARDS-814: Add a flatten json processor for forms

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,137 @@
+@Component(immediate = true)
+public class FlattenFormProcessor implements ResourceJsonProcessor
+{
+    // Saves all children with type card:Answer of root node
```
```suggestion
    // Saves all children with type cards:Answer of root node
```

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,137 @@
+{
+    // Saves all children with type card:Answer of root node
+    private ThreadLocal<Map<String, JsonObject>> childrenJsons = ThreadLocal.withInitial(HashMap::new);
+    // Saves all children with type card:Answer of specific AnswerSection node
```
```suggestion
    // Saves all children with type cards:Answer of specific AnswerSection node
```

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,137 @@
+    private ThreadLocal<Map<String, JsonObject>> childrenJsons = ThreadLocal.withInitial(HashMap::new);
+    // Saves all children with type card:Answer of specific AnswerSection node
+    private ThreadLocal<JsonObjectBuilder> jsonSection = ThreadLocal.withInitial(Json::createObjectBuilder);
+    // Saves all AnswerSection nodes ids with all children with type card:Answer that belong them
```
```suggestion
    // Saves all AnswerSection nodes ids with all children with type cards:Answer that belong them
```
Also, please provide an example of what is being mapped to what. For example, are these AnswerSection node identifiers (JCR Path or `jcr:uuid`?) being mapped to a _list_ or a _hash map_ of child `cards:Answer` nodes?

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,137 @@
+        return resource.isResourceType("cards/Form");
+    }
+
+    // Each time it is invoked, it is passed a JCR node and if that JCR node is a cards:Answer node, this node is
```
```suggestion
    // Each time this method is invoked, it is passed a JCR node and if that JCR node is a cards:Answer node, this node is
```

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,137 @@
+
+    // Each time it is invoked, it is passed a JCR node and if that JCR node is a cards:Answer node, this node is
+    // serialized and stored to a (this.childrenJsons) ThreadLocal object.
+    // We need to proceed and change representation just Answer which is saved in Answer Section, that is why in case
+    // of cards:Answer and cards:AnswerSection, otherwise the input JsonValue is simply passed-through.
```
```suggestion
    // We need to change the serialized data representation only for cards:Answer and cards:AnswerSection nodes, which
    // is why nodes that are not cards:Answer or cards:AnswerSection, are simply passed-through.
```

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,137 @@
+                if (child.isNodeType("cards:Answer") && this.childrenJsons.get().containsKey(childId)) {
+                    final JsonObject childJson = this.childrenJsons.get().get(childId);
+                    if (node.isNodeType("cards:AnswerSection")) {
+                        this.jsonSection.get().add(idWithoutParent, childJson);
+                        this.jsonSectionMap.get().put(node.getIdentifier(), this.jsonSection.get());
```
Add some comments describing what's going on here. Why are we modifying `this.jsonSection` and `this.jsonSectionMap`?

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,137 @@
+                        this.jsonSection.get().add(idWithoutParent, childJson);
+                        this.jsonSectionMap.get().put(node.getIdentifier(), this.jsonSection.get());
+                    } else {
+                        json.add(idWithoutParent, childJson);
```
Add a comment describing why this data can be added to the serialized data output stream.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,137 @@
+                    this.childrenJsons.get().remove(childId);
+                } else if (child.isNodeType("cards:AnswerSection") && this.jsonSectionMap.get().containsKey(childId)) {
+                    if (node.isNodeType("cards:AnswerSection")) {
+                        this.jsonSectionMap.get().put(node.getIdentifier(), this.jsonSectionMap.get().get(childId));
```
Add some comments describing what's going on here. Why is `this.jsonSectionMap` being modified?

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,137 @@
+                    } else {
+                        json.add(idWithoutParent, childJson);
+                    }
+                    this.childrenJsons.get().remove(childId);
```
Add a comment describing why `childId` is being removed from `this.childrenJsons`.

IntegralProgrammer commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/FlattenFormProcessor.java:
```diff
@@ -0,0 +1,137 @@
+                    if (node.isNodeType("cards:AnswerSection")) {
+                        this.jsonSectionMap.get().put(node.getIdentifier(), this.jsonSectionMap.get().get(childId));
+                    } else {
+                        json.addAll(this.jsonSectionMap.get().get(childId));
+                        this.jsonSectionMap.get().remove(childId);
```
Add some comments describing what's going on here. Why can `childId` be removed from `this.jsonSectionMap` and added to the serialized data output stream.

IntegralProgrammer reviewed (changes requested):
I tested this PR once again using the _Double Nested Section Test_ Form as well as another test Form that I manually created to have an Answer in 4 levels deep of AnswerSections. Both tests worked as expected! :+1: 

My comments/requested changes are mainly about commenting and documentation. Implementing these changes will make the code easier to understand and maintain.

sofi2002sofi committed:
CARDS-814: Add a flatten json processor for forms

add explaining comments, fix the method's logic in case of several Answer children in an AnswerSection

sdumitriu committed:
CARDS-814: Add a flatten json processor for forms
Simplify the code
Improve comments

sashaandjic reviewed (approved):
tested

IntegralProgrammer reviewed (approved):
Code looks good. Tested with the _Double Nested Section Test_ Form and a custom made 5-levels-deep nested section test Form. All works as expected. Approved :heavy_check_mark:.
